docling_eval/evaluators/ocr_evaluator.py

```diff
@@ -0,0 +1,152 @@
+import glob
+import json
+import logging
+import os
+import statistics
+from pathlib import Path
+from typing import Any, Dict, Generic, List, Optional, TypeVar
+
+import evaluate
+import pandas as pd
+from datasets import Dataset, load_dataset
+from docling_core.types.doc.document import DoclingDocument
+from pydantic import BaseModel
+from tqdm import tqdm
+
+from docling_eval.datamodels.dataset_record import DatasetRecordWithPrediction
+from docling_eval.datamodels.types import BenchMarkColumns, PredictionFormats
+from docling_eval.evaluators.base_evaluator import BaseEvaluator
+
+logging.basicConfig(
+    level=logging.DEBUG,
+    format="%(asctime)s - %(levelname)s - %(message)s",
+)
+
+_log = logging.getLogger(__name__)
+
+
+class PageEvaluation(BaseModel):
+    doc_id: str
+    gd_text: str
+    extracted_text: str
+    cer: float
+    char_accuracy: float
+
+
+class DatasetOcrEvaluation(BaseModel):
+    evaluations: List[PageEvaluation]
+    mean_character_accuracy: float
+
+
+class OCREvaluator(BaseEvaluator):
+    """Evaluator for OCR tasks that computes Character Accuracy"""
+
+    def __init__(
+        self,
+        intermediate_evaluations_path: Optional[Path] = None,
+        prediction_sources: List[PredictionFormats] = [
+            PredictionFormats.DOCLING_DOCUMENT
+        ],
+    ):
+        """Initialize the OCR evaluator"""
+        super().__init__(
+            intermediate_evaluations_path=intermediate_evaluations_path,
+            prediction_sources=prediction_sources,
+            supported_prediction_formats=[PredictionFormats.DOCLING_DOCUMENT],
+        )
+        # Load the CER evaluation metric
+        # https://huggingface.co/spaces/evaluate-metric/cer
+        self._cer_eval = evaluate.load("cer")
+
+    def __call__(
+        self,
+        ds_path: Path,
+        split: str = "test",
+    ) -> DatasetOcrEvaluation:
+
+        _log.info("Loading the split '%s' from: '%s'", split, ds_path)
+        split_path = str(ds_path / split / "*.parquet")
+        split_files = glob.glob(split_path)
+        _log.info("Files: %s", split_files)
+        ds = load_dataset("parquet", data_files={split: split_files})
+        _log.info("Overview of dataset: %s", ds)
+
+        # Select the split
+        ds_selection: Dataset = ds[split]
+
+        text_evaluations_list = []
+        char_accuracy_list = []
+
+        for i, data in tqdm(
+            enumerate(ds_selection),
+            desc="Evaluating OCR",
+            ncols=120,
+            total=len(ds_selection),
+        ):
+            data_record = DatasetRecordWithPrediction.model_validate(data)
+            doc_id = data_record.doc_id
+            true_doc_dict = data_record.ground_truth_doc
+            pred_doc_dict = data_record.predicted_doc
+
```

Comment:
These two are `DoclingDocument` instances, the `_dict` in the variable name is a bit misleading.